polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -825,47 +906,62 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				Ok(())
 			},
 			ExportMessage { network, destination, xcm } => {
-				// The actual message sent to the bridge for forwarding is prepended with
-				// `UniversalOrigin` and `DescendOrigin` in order to ensure that the message is
-				// executed with this Origin.
-				//
-				// Prepend the desired message with instructions which effectively rewrite the
-				// origin.
-				//
-				// This only works because the remote chain empowers the bridge
-				// to speak for the local network.
-				let origin = self.context.origin.ok_or(XcmError::BadOrigin)?;
-				let universal_source = Config::UniversalLocation::get()
-					.within_global(origin)
-					.map_err(|()| XcmError::Unanchored)?;
-				let hash = (self.origin_ref(), &destination).using_encoded(blake2_128);
-				let channel = u32::decode(&mut hash.as_ref()).unwrap_or(0);
-				// Hash identifies the lane on the exporter which we use. We use the pairwise
-				// combination of the origin and destination to ensure origin/destination pairs will
-				// generally have their own lanes.
-				let (ticket, fee) = validate_export::<Config::MessageExporter>(
-					network,
-					channel,
-					universal_source,
-					destination,
-					xcm,
-				)?;
-				self.take_fee(fee, FeeReason::Export(network))?;
-				Config::MessageExporter::deliver(ticket)?;
-				Ok(())
+				let old_holding = self.holding.clone();
+				let result = Config::TransactionalProcessor::process(|| -> Result<(), XcmError> {
+					// The actual message sent to the bridge for forwarding is prepended with
+					// `UniversalOrigin` and `DescendOrigin` in order to ensure that the message is
+					// executed with this Origin.
+					//
+					// Prepend the desired message with instructions which effectively rewrite the
+					// origin.
+					//
+					// This only works because the remote chain empowers the bridge
+					// to speak for the local network.
+					let origin = self.context.origin.ok_or(XcmError::BadOrigin)?;
+					let universal_source = Config::UniversalLocation::get()
+						.within_global(origin)
+						.map_err(|()| XcmError::Unanchored)?;
+					let hash = (self.origin_ref(), &destination).using_encoded(blake2_128);
+					let channel = u32::decode(&mut hash.as_ref()).unwrap_or(0);
+					// Hash identifies the lane on the exporter which we use. We use the pairwise
+					// combination of the origin and destination to ensure origin/destination pairs
+					// will generally have their own lanes.
+					let (ticket, fee) = validate_export::<Config::MessageExporter>(
+						network,
+						channel,
+						universal_source,
+						destination,
+						xcm,
+					)?;
+					self.take_fee(fee, FeeReason::Export(network))?;
+					Config::MessageExporter::deliver(ticket)?;
```

Comment:
Nothing has changed prior to `take_fee`, and `deliver` can only return an error if the `ticket` has become invalid (which it can't have done). Therefore transactionality isn't needed.